src/advanced/JupyterNotebook/index.tsx

```diff
@@ -0,0 +1,322 @@
+    );
+};
+
+// -- renders a cell of type 'error'
+const JupyterNotebookOutputErrorCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    executionCount,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`In [${executionCount}]`}
+            bodyColumnContent={
+                <ReactMarkdown
+                    children={
+                        '**[' +
+                        output.evalue +
+                        ']**\n```\n ' +
+                        sourceLines(output.traceback).replace(
+                            /[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, // strip out ansi control sequences common in output traces
+                            ''
+                        ) +
+                        '\n```'
+                    }
+                />
+            }
+            bodyColumnStyle={{
+                padding: '0px 8px 0px 8px',
+                margin: '0px 0px',
+                border: '1px solid #AAA',
+                backgroundColor: '#FDD',
+            }}
+        />
+    );
+};
+
+// -- renders various data cell types (html, png, text)
+const JupyterNotebookOutputDataCell: React.FC<JupyterNotebookOutputCellProps> = ({
+    cell,
+    executionCount,
+    output,
+    language,
+}) => {
+    return (
+        <JupyterNotebookCellRenderTemplate
+            idColumnContent={`Out [${executionCount}]`}
+            bodyColumnContent={
+                <Container
+                    style={{
+                        marginBottom: '0px',
+                        boxShadow: 'none',
+                    }}
+                >
+                    {output.data['text/html'] ? (
+                        parse(sourceLines(output.data['text/html']), { trim: true })
+                    ) : output.data['image/png'] ? (
+                        <img
+                            src={'data:image/png;base64,' + output.data['image/png'].replace(/\n/g, '')}
+                            alt={'notebook diagram'}
+                        />
+                    ) : output.data['text/plain'] ? (
+                        <ReactMarkdown children={'```\n' + sourceLines(output.data['text/plain']) + '\n```'} />
+                    ) : (
+                        <b>unknown</b>
+                    )}
+                </Container>
+            }
+            bodyColumnStyle={{ padding: '0px 0px 0px 0px', margin: '0px 0px', border: '1px solid #AAA' }}
+        />
+    );
+};
+
+// -- helper class used in syntax highlighting the code
+const CodeBlock: CodeComponent = ({ inline = false, className, children }) => {
+    const match = /language-(\w+)/.exec(className || '');
+    const codeLanguage = useMemo(() => {
+        const langs = ['python', 'java', 'javascript', 'c++', 'typescript', 'objective-c', 'json'];
+        return langs.find((lang) => match && match[1] && match[1].startsWith(lang)) || 'javascript';
```

Comment:
```js
return langs.find((lang) => match?.[1]?.startsWith(lang)) || 'javascript';
```

Reply:
will change, but I'm concerned that maybe a little terse